src/include/denc.h

```diff
@@ -1399,34 +1589,48 @@ inline typename std::enable_if<traits::supported &&
   traits::encode(o, a, features);
 }
 
-template<typename T, typename traits=denc_traits<T>>
+template<typename T,
+	 typename traits=denc_traits<T>>
 inline typename std::enable_if<traits::supported &&
-			       !traits::featured>::type decode(
+			       traits::with_dencv>::type decode(
   T& o,
   bufferlist::iterator& p)
 {
   if (p.end())
     throw buffer::end_of_buffer();
-  // ensure we get a contigous buffer... until the end of the
-  // bufferlist.  we don't really know how much we'll need here,
-  // unfortunately.  hopefully it is already contiguous and we're just
-  // bumping the raw ref and initializing the ptr tmp fields.
-  bufferptr tmp;
-  bufferlist::iterator t = p;
-  t.copy_shallow(p.get_bl().length() - p.get_off(), tmp);
-  auto cp = tmp.begin();
-  traits::decode(o, cp);
-  p.advance((ssize_t)cp.get_offset());
+  auto bl = p.get_bl();
+  const auto remaining = bl.length() - p.get_off();
+  // it is expensive to rebuild a contigous buffer and drop it, so avoid this.
+  if (p.get_current_ptr().get_raw() != bl.back().get_raw() &&
+      remaining > CEPH_PAGE_SIZE) {
+    traits::decode(o, p);
+  } else {
+    // ensure we get a contigous buffer... until the end of the
+    // bufferlist.  we don't really know how much we'll need here,
+    // unfortunately.  hopefully it is already contiguous and we're just
+    // bumping the raw ref and initializing the ptr tmp fields.
+    bufferptr tmp;
+    bufferlist::iterator t = p;
+    t.copy_shallow(remaining, tmp);
+    auto cp = tmp.begin();
+    traits::decode(o, cp);
+    p.advance((ssize_t)cp.get_offset());
```

Comment:
> if it can be read by multiple threads in parallel, we are in the risk of racing.

true

> i think, if the caller is concerned about the cost of repeatedly flatten the decoded buffer, it should cache the decoded object instead keeping the bufferlist around

by later decodes, i'm referring to the rest of the elements being decoded from this bufferlist. say you start with a bufferlist with 8 segments of 8 bytes each `[8][8][8][8][8][8][8][8]`, and you're decoding four 16-byte objects. the first decode flattens the bufferlist into a new ptr `[64]` and decodes from that. the next decode still sees `[8][8][8][8][8][8]` left, and has to flatten again and decode from a ptr `[48]`. in the end, you're allocating and copying buffers for each object separately